### ISSUE TITLE
Problems training tiny YOLOv3

### ISSUE THREAD

**Issue by nikolaevra** [AlexeyAB/darknet#3707]
I'm currently using this command to train my tiny yolo on [WIDER dataset](http://shuoyang1213.me/WIDERFACE/) for face detection (1 class):
`./darknet detector train data/obj.data yolov3-tiny-obj.cfg yolov3-tiny.conv.15 -map`
But, I'm not sure if this is training only the head of the network or the entire network (including the darknet feature extractor)

Also, I'm currently at 20,000th iteration and my mAP looks kinda bad as well
```
 (next mAP calculation at 21100 iterations) 
 Last accuracy mAP@0.5 = 27.22 %, best = 27.22 % 
 20950: 4.561118, 4.670623 avg loss, 0.001000 rate, 2.471459 seconds, 1340800 images
```
As a note, I didn't use the `Yolo_mark` tool to label images, I just converted the ground truth labels provided in the WIDER FACE dataset. But I did double-check everything to make sure my conversions were the right format.

Can someone explain to me what the command above does, and what command is used to train the entire network?

Also, is it normal to have such low mAP by the 20,000th iteration?

Thanks!

**Comment by nikolaevra** [comment]
For context, here is what my `.cfg` file looks like
```
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 500200
policy=steps
steps=400000,450000
scales=.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear



[yolo]
mask = 3,4,5
anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
classes=1
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 8

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear

[yolo]
mask = 0,1,2
anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
classes=1
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1
```

**Comment by lsd1994** [comment]
For only one class, the mAP is not satisfied.
You can read this to improve
https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

**Comment by nikolaevra** [comment]
> For only one class, the mAP is not satisfied.

What does it mean? 

Does it mean that It's not enough to use the mAP to evaluate the model when training for one class only? If that is the case, what should be my target loss to stop training? My loss is currently around 4 and the lowest I've seen it was 3.7

**Comment by nikolaevra** [comment]
After browsing through the dataset a little more, there is a number of really small bounding boxes. I read through https://github.com/AlexeyAB/darknet#how-to-improve-object-detection and tried training the `yolov3-tiny_3l.cfg` model as well. The mAP went up to around `44%`, which is great! However, when I started to visually evaluate the model outputs, they are very inaccurate. Here are some examples.

![face_example_outputs](https://user-images.githubusercontent.com/19254121/62485862-e0b7f480-b78b-11e9-92ec-372b6940c9ea.png)

From these examples, It looks like the bounding boxes are too small for the face and only detect a part of the face, rather than the whole face

Some of the things I'm thinking of trying to improve accuracy:
1. Train even longer? I noticed that the loss started to increase at some point from 3.7 to 4.2+, that's why I stopped training in the first place. But then the question is again, what loss should I be looking for, since mAP is [not a factor](https://github.com/AlexeyAB/darknet/issues/3707#issuecomment-518055535)?
2. I could remove all of the images that have tiny label boxes and only keep the ones that have larger boxes. (since in production environment, I don't need to detect small faces). The dataset is plenty large, so I can afford to remove a bunch of pictures.

What do you guys think? What would be the recommended solution in this case to improve the accuracy of the bounding box sizes?

**Comment by lsd1994** [comment]
> > For only one class, the mAP is not satisfied.
> 
> What does it mean?
> 
> Does it mean that It's not enough to use the mAP to evaluate the model when training for one class only? If that is the case, what should be my target loss to stop training? My loss is currently around 4 and the lowest I've seen it was 3.7

I mean mAP=0.27, even 0.44 is not good for detecting only one class. mAP is enough to evaluate.
I think you can further reduce the loss since 3.7 is still a high loss. If your labels are correct, remove small object may not help.

**Comment by ntd94** [comment]
Did you recalculate anchors for your dataset?

**Comment by nikolaevra** [comment]
> > > For only one class, the mAP is not satisfied.
> > 
> > 
> > What does it mean?
> > Does it mean that It's not enough to use the mAP to evaluate the model when training for one class only? If that is the case, what should be my target loss to stop training? My loss is currently around 4 and the lowest I've seen it was 3.7
> 
> I mean mAP=0.27, even 0.44 is not good for detecting only one class. mAP is enough to evaluate.
> I think you can further reduce the loss since 3.7 is still a high loss. If your labels are correct, remove small objects may not help.

So when you are saying that I can further decrease the loss, you are recommending to train the model even longer? So far the longest I've trained is around 30,000 iterations.

**Comment by nikolaevra** [comment]
> Did you recalculate anchors for your dataset?

Yes, I did

**Comment by nikolaevra** [comment]
UPDATE: I started to get decent results. The reason I think it started to work is because I retrained the model from scratch.

I ran the `/darknet detector train obj.data tiny-yolo_obj.cfg -clear -map` command to train the model from scratch. This actually produces much better results while still being only at `~31%` mAP. The loss is still quite high at `~3.9`. But at least, the predictions are looking at lot better.
![face_example_outputs2](https://user-images.githubusercontent.com/19254121/62555249-9b093380-b840-11e9-828c-86750a871781.png)

**Comment by nikolaevra** [comment]
One last question is how do I finetune the model. The WIDER dataset that I used for training is not very representative of my production data. I want to fine-tune the model on my smaller hand-labelled production dataset. What command do I use for that?

I'm thinking that I can use
1. `darknet.exe partial cfg/yolov3-tiny.cfg yolov3-tiny_obj_best.weights yolov3-tiny.conv.15 15`
to extract the weights for the feature extractor, and then
2. `/darknet detector train obj_finetune.data yolo-tiny_obj,cfg yolov3-tiny.conv.15` 
to train on fine-tune data

The question I have is what does `15` mean in the command (1)?

**Comment by lsd1994** [comment]
The result is pretty good. It seems your dataset is difficult so 3.9 loss is also acceptable.
You can also try some state-of-the-art face detection model, as yolo is an object detection model.

**Comment by nikolaevra** [comment]
> The result is pretty good. It seems your dataset is difficult so 3.9 loss is also acceptable.
> You can also try some state-of-the-art face detection model, as yolo is an object detection model.

Do you have any recommendations for the models I should check out?

**Comment by lsd1994** [comment]
> > The result is pretty good. It seems your dataset is difficult so 3.9 loss is also acceptable.
> > You can also try some state-of-the-art face detection model, as yolo is an object detection model.
> 
> Do you have any recommendations for the models I should check out?

I‘m not familiar in face detection...

**Comment by usamahjundia** [comment]
@nikolaevra hi, did you eventually managed to fine tune the widerface trained model? how did you do it?
